skimage/util/dtype.py

```diff
@@ -2,28 +2,29 @@
 import numpy as np
 from warnings import warn
 
+
 __all__ = ['img_as_float32', 'img_as_float64', 'img_as_float',
            'img_as_int', 'img_as_uint', 'img_as_ubyte',
            'img_as_bool', 'dtype_limits']
 
+# For integers Numpy uses `_integer_types` basis internally, and builds a leaky
+# `np.XintYY` abstraction on top of it. This leads to situations when, for
+# example, there are two np.Xint64 dtypes with the same attributes but
+# different object references. In order to avoid any potential issues,
+# we use the basis dtypes here. For more information, see:
+# - https://github.com/scikit-image/scikit-image/issues/3043
```

Comment:
@soupault I'd love a table with some reasonable guesses about what these correspond to, e.g.

```python
_integer_types = (
    np.byte,  # int8
    np.short,  # int16
    np.intc,  # platform-specific, typically int32 or int64
    [...]
```

I'm not sure whether this is feasible or reasonable, but even small hints would be nice.

Incidentally, is this list guaranteed to cover all integer types? e.g. is can `intc`, `int_`, and `longlong` all point to 64bit, and then you end up not being able to convert 32-bit ints?

Reply:
>a table with some reasonable guesses

Sure. I'll come up with the update.

>is this list guaranteed to cover all integer types?

I've been convenced by @eric-wieser that this list is indeed complete (as of 03.05.2018 :smile:).

Reply:
The list of types matches 1:1 with the native integer types in C. The situation you suggest is still possible though - but it means the platform does not support 32-bit integers, and that np.uint32 does not exist at all.

Reply:
@jni see the update. The info is taken from the table in http://www.unix.org/whitepapers/64bit.html .

Reply:
@soupault that link should be in the comment! =)

Travis is totally hosed now that the new SciPy is out. 🤦‍♂️

Reply:
I feel like my PRs undergo 9 circles of review :wink: . Done for the link.

Reply:
@soupault I learned to review from the best/most annoying! (👋 @stefanv) =P

In seriousness, I did learn from @stefanv that if I don't understand a PR first shot, then likely a maintainer in the future will also not understand. In these cases I try to force as much information into the comments as possible. =) Thanks for hanging in there!

Reply:
You'll thank me one day :P

Honestly, though, even I sometimes get annoyed at my code undergoing multiple rounds of review; I think it's only human to eventually start reading it as personal criticism.

One way we can improve the contributor experience is by pushing highly specific / nitpicky changes directly to the PR.  It feels kind of strange to do that—it's "their" code after all—but it helps to keep in mind that once it's merged it'll be everybody's maintenance burden anyway!

Reply:
Also, a big +1 to always erring on the over-descriptive side; it's pretty easy to remove comments later on, but often quite hard to go back and do the research to expand them.

I'd also like to see us gradually expand the error messages in skimage to provide users with much more guidance, even if it takes a small document to explain it.